### PR TITLE
expression: fix the arg verification for json functions.

### LINKED ISSUE #54029: Using JSON_TYPE on non-JSON values should return an error
### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE t1(id INT PRIMARY KEY, d1 DATE, d2 DATETIME, t1 TIME, t2 TIMESTAMP, b1 BIT, b2 BINARY);
INSERT INTO t1 VALUES (1, '2024-06-14', '2024-06-14 09:37:00', '09:37:00', '2024-06-14 09:37:00', b'0', 0x41);
SELECT JSON_TYPE(d1),JSON_TYPE(d2),JSON_TYPE(t1),JSON_TYPE(t2),JSON_TYPE(b1),JSON_TYPE(b2) FROM t1;
```

### 2. What did you expect to see? (Required)

```
mysql-8.4.0> CREATE TABLE t1(id INT PRIMARY KEY, d1 DATE, d2 DATETIME, t1 TIME, t2 TIMESTAMP, b1 BIT, b2 BINARY);
Query OK, 0 rows affected (0.02 sec)

mysql-8.4.0> INSERT INTO t1 VALUES (1, '2024-06-14', '2024-06-14 09:37:00', '09:37:00', '2024-06-14 09:37:00', b'0', 0x41);
Query OK, 1 row affected (0.01 sec)

mysql-8.4.0> SELECT JSON_TYPE(d1),JSON_TYPE(d2),JSON_TYPE(t1),JSON_TYPE(t2),JSON_TYPE(b1),JSON_TYPE(b2) FROM t1;
ERROR 3146 (22032): Invalid data type for JSON data in argument 1 to function json_type; a JSON string or JSON type is required.
```

### 3. What did you see instead (Required)

```
mysql-8.0.11-TiDB-v8.1.0> CREATE TABLE t1(id INT PRIMARY KEY, d1 DATE, d2 DATETIME, t1 TIME, t2 TIMESTAMP, b1 BIT, b2 BINARY);
Query OK, 0 rows affected (0.11 sec)

mysql-8.0.11-TiDB-v8.1.0> INSERT INTO t1 VALUES (1, '2024-06-14', '2024-06-14 09:37:00', '09:37:00', '2024-06-14 09:37:00', b'0', 0x41);
Query OK, 1 row affected (0.01 sec)

mysql-8.0.11-TiDB-v8.1.0> SELECT JSON_TYPE(d1),JSON_TYPE(d2),JSON_TYPE(t1),JSON_TYPE(t2),JSON_TYPE(b1),JSON_TYPE(b2) FROM t1;
+---------------+---------------+---------------+---------------+------------------+---------------+
| JSON_TYPE(d1) | JSON_TYPE(d2) | JSON_TYPE(t1) | JSON_TYPE(t2) | JSON_TYPE(b1)    | JSON_TYPE(b2) |
+---------------+---------------+---------------+---------------+------------------+---------------+
| DATE          | DATETIME      | TIME          | DATETIME      | UNSIGNED INTEGER | BLOB          |
+---------------+---------------+---------------+---------------+------------------+---------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v8.1.0
Edition: Community
Git Commit Hash: 945d07c5d5c7a1ae212f6013adfb187f2de24b23
Git Branch: HEAD
UTC Build Time: 2024-05-21 03:51:57
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### LINKED ISSUE #54044: Error 3144 (HY000): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
## Bug Report
the bug is same as MySQL, [link](https://bugs.mysql.com/bug.php?id=93052)

### 1. Minimal reproduce step (Required)
when use gorm in golang.
```
storage.DB().Model(filter).Where(filter).Where("JSON_CONTAINS(labels,JSON_OBJECT(?,?))", "flag", true).Find(&list).Error
```
### 2. What did you expect to see? (Required)
The query result is displayed normally

### 3. What did you see instead (Required)
```
Error 3144 (HY000): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
```

If concatenated into an SQL statement, it runs properly.

### 4. What is your TiDB version? (Required)
Release Version: v6.5.0
Edition: Community
Git Commit Hash: 706c3fa3c526cdba5b3e9f066b1a568fb96c56e3
Git Branch: heads/refs/tags/v6.5.0
UTC Build Time: 2022-12-27 03:50:44
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #54029, close #54044

Problem Summary:

For some json functions, the arg verification is not implemented correctly:

1. For `json_type`, it didn't verify the type of args.
2. For many other json functions, they didn't verify the charset of the args.

For `PREPARE`, the param marker is set to binary string type, which is not expected. This PR will give it a default collation. It will not affect the `EXECUTE`, because when it's executing a statement, it'll infer the collation according to the datum (whose collation is set according to the collation settings of the connection).

### What changed and how does it work?

1. Add a unified function to verify the args of json functions.
2. Modify the collation of param marker from binary to default collation.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

### Release note

```release-note
Fix the issue that `json_type` can accept non-JSON arguments.
Fix the issue that some json functions accept binary string arguments.
Fix the issue that `PREPARE` will fail if the statement contains some json related functions.
```

### CODE REVIEW

**Comment by YangKeao** [comment]
/hold

Also consider https://github.com/pingcap/tidb/issues/54044#issuecomment-2180954644. Maybe it's not good to verify args type/charset in `getFunction`, because it'll be called in `PREPARE` stage.

File: tests/integrationtest/r/expression/json.result [hunk 1]
```diff
@@ -727,7 +727,7 @@ Error 3146 (22032): Invalid data type for JSON data in argument 1 to function js
 SELECT JSON_OVERLAPS('{}', 123);
 Error 3146 (22032): Invalid data type for JSON data in argument 2 to function json_overlaps; a JSON string or JSON type is required.
 SELECT 'abc' MEMBER OF(123);
-Error 3146 (22032): Invalid data type for JSON data in argument 2 to function member of; a JSON string or JSON type is required.
+Error 3146 (22032): Invalid data type for JSON data in argument 2 to function json_memberof; a JSON string or JSON type is required.
```
Comment (dveeden) [inline tests/integrationtest/r/expression/json.result:730]: Looks like this error need to be fixed.

MySQL v8.4.0:
```
mysql-8.4.0> SELECT 'abc' MEMBER OF (123);
ERROR 3146 (22032): Invalid data type for JSON data in argument 2 to function member of; a JSON string or JSON type is required.
```

This is a weird function syntax wise...

File: pkg/expression/builtin_json.go [hunk 2]
```diff
@@ -765,10 +797,7 @@ func (c *jsonMemberOfFunctionClass) verifyArgs(ctx EvalContext, args []Expressio
 	if err := c.baseFunctionClass.verifyArgs(args); err != nil {
 		return err
 	}
-	if evalType := args[1].GetType(ctx).EvalType(); evalType != types.ETJson && evalType != types.ETString {
-		return ErrInvalidTypeForJSON.GenWithStackByArgs(2, "member of")
-	}
-	return nil
+	return verifyJSONArgsType(ctx, c.funcName, true, args, 1)
```
Comment (dveeden) [inline pkg/expression/builtin_json.go:800]:
```suggestion
	return verifyJSONArgsType(ctx, "member of", true, args, 1)
```

**Comment by dveeden** [comment]
@YangKeao can you update this PR?

File: tests/integrationtest/r/expression/json.result (hunk 1 above)
Comment (YangKeao) [reply tests/integrationtest/r/expression/json.result]: Done.

File: pkg/expression/builtin_json.go (hunk 2 above)
Comment (YangKeao) [reply pkg/expression/builtin_json.go]: Done.

**Comment by YangKeao** [comment]
> @YangKeao can you update this PR?

Done. PTAL

File: pkg/types/field_type.go [hunk 3]
```diff
@@ -179,7 +179,7 @@ func SetTypeFlag(flag *uint, flagItem uint, on bool) {
 // InferParamTypeFromDatum is used for plan cache to infer the type of a parameter from its datum.
 func InferParamTypeFromDatum(d *Datum, tp *FieldType) {
 	InferParamTypeFromUnderlyingValue(d.GetValue(), tp)
-	if IsStringKind(d.k) {
+	if IsStringKind(d.k) || d.GetValue() == nil {
```
Comment (Defined2014) [inline pkg/types/field_type.go:182]: Can you explain why I need to set the charset again here when I have already set it in the `InferParamTypeFromUnderlyingValue` function?

**Review by windtalker — APPROVED** [review]
lgtm

**Review by Defined2014 — APPROVED** [review]

**Review by dveeden — CHANGES_REQUESTED** [review]
```sql
CREATE TABLE t1(id INT PRIMARY KEY, d1 DATE, d2 DATETIME, t1 TIME, t2 TIMESTAMP, b1 BIT, b2 BINARY);
INSERT INTO t1 VALUES (1, '2024-06-14', '2024-06-14 09:37:00', '09:37:00', '2024-06-14 09:37:00', b'0', 0x41);
SELECT JSON_TYPE(b2) FROM t1;
```

TiDB:
```
mysql-8.0.11-TiDB-v8.5.0-alpha-160-gce19496ae9-dirty> SELECT JSON_TYPE(b2) FROM t1;
ERROR 3144 (HY000): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
```

MySQL 9.1.0:
```
mysql-9.1.0> SELECT JSON_TYPE(b2) FROM t1;
ERROR 3144 (22032): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
```

The SQL State is 22032 for MySQL and HY000 for TiDB

**Comment by dveeden** [comment]
```patch
diff --git a/pkg/parser/mysql/errcode.go b/pkg/parser/mysql/errcode.go
index 05d5dc3e69..bfbd196138 100644
--- a/pkg/parser/mysql/errcode.go
+++ b/pkg/parser/mysql/errcode.go
@@ -898,6 +898,7 @@ const (
 	ErrInvalidJSONText                                       = 3140
 	ErrInvalidJSONTextInParam                                = 3141
 	ErrInvalidJSONPath                                       = 3143
+	ErrInvalidJSONCharset                                    = 3144
 	ErrInvalidTypeForJSON                                    = 3146
 	ErrInvalidJSONPathWildcard                               = 3149
 	ErrInvalidJSONContainsPathType                           = 3150
diff --git a/pkg/parser/mysql/errname.go b/pkg/parser/mysql/errname.go
index f757f825da..845c50a21b 100644
--- a/pkg/parser/mysql/errname.go
+++ b/pkg/parser/mysql/errname.go
@@ -907,6 +907,7 @@ var MySQLErrName = map[uint16]*ErrMessage{
 	ErrInvalidJSONText:                                       Message("Invalid JSON text: %-.192s", nil),
 	ErrInvalidJSONTextInParam:                                Message("Invalid JSON text in argument %d to function %s: \"%s\" at position %d.", nil),
 	ErrInvalidJSONPath:                                       Message("Invalid JSON path expression %s.", nil),
+	ErrInvalidJSONCharset:                                    Message("Cannot create a JSON value from a string with CHARACTER SET '%s'.", nil),
 	ErrInvalidTypeForJSON:                                    Message("Invalid data type for JSON data in argument %d to function %s; a JSON string or JSON type is required.", nil),
 	ErrInvalidJSONPathWildcard:                               Message("In this situation, path expressions may not contain the * and ** tokens or an array range.", nil),
 	ErrInvalidJSONContainsPathType:                           Message("The second argument can only be either 'one' or 'all'.", nil),
diff --git a/pkg/parser/mysql/state.go b/pkg/parser/mysql/state.go
index 2cbc6f1d2b..307965d088 100644
--- a/pkg/parser/mysql/state.go
+++ b/pkg/parser/mysql/state.go
@@ -254,6 +254,7 @@ var MySQLState = map[uint16]string{
 	ErrInvalidJSONText:                     "22032",
 	ErrInvalidJSONTextInParam:              "22032",
 	ErrInvalidJSONPath:                     "42000",
+	ErrInvalidJSONCharset:                  "22032",
 	ErrInvalidJSONData:                     "22032",
 	ErrInvalidJSONPathWildcard:             "42000",
 	ErrJSONUsedAsKey:                       "42000",
```

**Comment by YangKeao** [comment]
> ```sql
> CREATE TABLE t1(id INT PRIMARY KEY, d1 DATE, d2 DATETIME, t1 TIME, t2 TIMESTAMP, b1 BIT, b2 BINARY);
> INSERT INTO t1 VALUES (1, '2024-06-14', '2024-06-14 09:37:00', '09:37:00', '2024-06-14 09:37:00', b'0', 0x41);
> SELECT JSON_TYPE(b2) FROM t1;
> ```
> 
> TiDB:
> 
> ```
> mysql-8.0.11-TiDB-v8.5.0-alpha-160-gce19496ae9-dirty> SELECT JSON_TYPE(b2) FROM t1;
> ERROR 3144 (HY000): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
> ```
> 
> MySQL 9.1.0:
> 
> ```
> mysql-9.1.0> SELECT JSON_TYPE(b2) FROM t1;
> ERROR 3144 (22032): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
> ```
> 
> The SQL State is 22032 for MySQL and HY000 for TiDB


Nice catch! Fixed. PTAL @dveeden

**Review by dveeden — APPROVED** [review]

File: tests/integrationtest/r/expression/json.result [hunk 4]
```diff
@@ -771,3 +771,49 @@ insert into t values (NULL, NULL, NULL);
...
+SELECT JSON_TYPE(d2) FROM t1;
+Error 3146 (22032): Invalid data type for JSON data in argument 1 to function json_type; a JSON string or JSON type is required.
+SELECT JSON_TYPE(t1) FROM t1;
+Error 3146 (22032): Invalid data type for JSON data in argument 1 to function json_type; a JSON string or JSON type is required.
+SELECT JSON_TYPE(t2) FROM t1;
+Error 3146 (22032): Invalid data type for JSON data in argument 1 to function json_type; a JSON string or JSON type is required.
+SELECT JSON_TYPE(b1) FROM t1;
+Error 3146 (22032): Invalid data type for JSON data in argument 1 to function json_type; a JSON string or JSON type is required.
+SELECT JSON_TYPE(b2) FROM t1;
+Error 3144 (HY000): Cannot create a JSON value from a string with CHARACTER SET 'binary'.
```
Comment (Defined2014) [inline tests/integrationtest/r/expression/json.result:788]: The sql state seems not changed.

Comment (YangKeao) [reply tests/integrationtest/r/expression/json.result]: Forgot to update the test result. It's changed now :+1:

**Pull request merged**